test/test_ff_utils.py

```diff
@@ -471,28 +479,77 @@ def test_upsert_metadata(integrated_ff):
+        search_items = item_maker()[search_from:search_from + search_limit]
+        if parsed.path.endswith("/browse/"):
+            restype = 'Browse'
+        elif parsed.path.endswith("/search/"):
+            restype = 'Search'
+        else:
+            raise NotImplementedError("Need a better mock.")
+        print("mocked search", url)
+        if len(search_items) > 5:
+            print("yields %s [%s, ..., %s] # %s items"
+                  % (restype, search_items[0], search_items[-1], len(search_items)))
+        else:
+            print("yields %s %s # %s items" % (restype, search_items, len(search_items)))
+        return MockResponse(json={'@type': restype, '@graph': search_items})
+    return mocked_search
+
+
 @pytest.mark.parametrize('url', ['', 'to_become_full_url'])
-def test_search_metadata(integrated_ff, url):
-    from types import GeneratorType
-    if (url != ''):  # replace stub with actual url from integrated_ff
+def test_search_metadata_unit(integrated_ff, url):
+    with mock.patch.object(ff_utils, "authorized_request", make_mocked_search()):
+        check_search_metadata(integrated_ff, url)
+
+
+# @pytest.mark.integrated
+# @pytest.mark.flaky
+# @pytest.mark.parametrize('url', ['', 'to_become_full_url'])
+# def test_search_metadata_integrated(integrated_ff, url):
+#     check_search_metadata(integrated_ff, url)
+
+
+def check_search_metadata(integrated_ff, url):
+    if url != '':  # replace stub with actual url from integrated_ff
         url = integrated_ff['ff_key']['server'] + '/'
     search_res = ff_utils.search_metadata(url + 'search/?limit=all&type=File', key=integrated_ff['ff_key'])
     assert isinstance(search_res, list)
     # this will fail if items have not yet been indexed
     assert len(search_res) > 0
     # make sure uuids are unique
-    search_uuids = set([item['uuid'] for item in search_res])
-    assert len(search_uuids) == len(search_res)
+    check_duplicated_items_by_key('uuid', search_res, url=url, formatter=lambda x: json.dumps(x, indent=2))
+    # search_uuids = set([item['uuid'] for item in search_res])
+    # assert len(search_uuids) == len(search_res)
     search_res_slash = ff_utils.search_metadata(url + '/search/?limit=all&type=File', key=integrated_ff['ff_key'])
     assert isinstance(search_res_slash, list)
     assert len(search_res_slash) == len(search_res)
     # search with a limit
     search_res_limit = ff_utils.search_metadata(url + '/search/?limit=3&type=File', key=integrated_ff['ff_key'])
     assert len(search_res_limit) == 3
     # search with a limit from a certain entry
-    search_res_from_limit = ff_utils.search_metadata(url + '/search/?type=File&from=5&limit=53', key=integrated_ff['ff_key'])
+    search_res_from_limit = ff_utils.search_metadata(url + '/search/?type=File&from=5&limit=53',
+                                                     key=integrated_ff['ff_key'])
     assert len(search_res_from_limit) == 53
     # search with a filter
```

Comment:
Is this actually doing any filtering since it's using the mock result?

Reply:
It is not doing type filtering. It is giving back a different number of items based on what you do. The thing that's mocked is the authorized_request call. So if you run this test with suitable verbosity, you'll see this output (where the "..." and the "#" comments are not me abbreviating here, but actually the result of the output since I thought otherwise the output was overly verbose and impeded my debugging):
```
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=50
yields Search [{'uuid': '50'}, ..., {'uuid': '99'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=100
yields Search [{'uuid': '100'}, ..., {'uuid': '129'}] # 30 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=50
yields Search [{'uuid': '50'}, ..., {'uuid': '99'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=100
yields Search [{'uuid': '100'}, ..., {'uuid': '129'}] # 30 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?type=File&from=5&limit=50&sort=-date_created
yields Search [{'uuid': '5'}, ..., {'uuid': '54'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?type=File&from=55&limit=50&sort=-date_created
yields Search [{'uuid': '55'}, ..., {'uuid': '104'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&file_type=reads&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&file_type=reads&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/browse/?limit=50&type=File&sort=-date_created&from=0
yields Browse [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
PASSED   [100%]mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=50
yields Search [{'uuid': '50'}, ..., {'uuid': '99'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com/search/?limit=50&type=File&sort=-date_created&from=100
yields Search [{'uuid': '100'}, ..., {'uuid': '129'}] # 30 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com//search/?limit=50&type=File&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com//search/?limit=50&type=File&sort=-date_created&from=50
yields Search [{'uuid': '50'}, ..., {'uuid': '99'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com//search/?limit=50&type=File&sort=-date_created&from=100
yields Search [{'uuid': '100'}, ..., {'uuid': '129'}] # 30 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com//search/?limit=50&type=File&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com//search/?type=File&from=5&limit=50&sort=-date_created
yields Search [{'uuid': '5'}, ..., {'uuid': '54'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com//search/?type=File&from=55&limit=50&sort=-date_created
yields Search [{'uuid': '55'}, ..., {'uuid': '104'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com//search/?limit=50&type=File&file_type=reads&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com//search/?limit=50&type=File&file_type=reads&sort=-date_created&from=0
yields Search [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
mocked search http://fourfront-mastertest.9wzadzju3p.us-east-1.elasticbeanstalk.com//browse/?limit=50&type=File&sort=-date_created&from=0
yields Browse [{'uuid': '0'}, ..., {'uuid': '49'}] # 50 items
```
I didn't see the test testing for any additional kind of filtering. I was trying to convert the test from an integration test to a unit test. The assumption in unit testing is that you only test the new layer of functionality and that the underlying stuff (in this case the search endpoint) is tested already elsewhere. So instead this just checks that the arguments are always what it expects so that if you start asking for different values, it's easier to see that the mock needs to be extended. Does that make sense? Is there some additional functionality that you don't think is tested?

A separate and reasonable question you might be asking is: Should we also have a set of integration tests or system tests that test the search functionality from the outside. The answer to that question might be yes. If you mean this to actually be a test of /search/ itself, that may be important. But that seems a separate thing from testing this endpoint.

Reply:
And there's a more subtle question: Really this functionality seems to stitch together the result of multiple paginated calls. As far as I can tell, if ES (due to indexing) is changing underneath, it won't get that right.  You will sometimes omit results. That problem is unsolvable probably. You can't know they are coming. But sometimes you'll get duplications. In that case, you could just recognize that it's an artifact of something getting introduced. It'd be expensive to back up and redo the calls, but minimally it might be good to make a pass through and de-dup the result rather than returning the same item more than once.

Reply:
I understand what a unit test is and what you are testing and my comment is just pointing out that the additional search with the file_type filter can be removed as it's really not doing anything.

Reply:
Well, when I reenable the part that's the integration test (which I plan to do one I have the integration and unit tests more fully separated) then it will be needed to keep the test small. So I think it's OK to leave, but I take your point. (Also, it somewhat has the vague purpose of making the point that we can't just look for some constant set of query args because there will be others littered in there.)